src/lib/homestore/heap_chunk_selector.cpp

```diff
@@ -0,0 +1,116 @@
+m_pdev_avalable_blk_map(PdevAvalableBlkMap(pdev_avalable_blk_map_initial_size)),
+m_chunks(ChunkMap(chunk_initial_num)) {}
+
+
+//add_chunk may be called in Homeobject(when sealing a shard) or in Homestore(when adding chunks to vdev) concurently
+void HeapChunkSelector::add_chunk(csharedChunk& chunk) {
+    VChunk vchunk(chunk);
+    auto&& pdevID = vchunk.get_pdev_id();
+    auto&& [it, _] = m_pdev_heap_map.emplace(pdevID, std::make_shared<std::pair<std::mutex, VChunkHeap>>());
+    const auto& available_blk_num = vchunk.available_blks();
+    auto&& [ignore, happened] = m_pdev_avalable_blk_map.emplace(pdevID, available_blk_num);
+
+    if(!happened) {
+        auto&& avalableBlkCounter = m_pdev_avalable_blk_map.find(pdevID)->second;
+        avalableBlkCounter.fetch_add(available_blk_num);
+    }
+    
+    m_chunks.emplace(vchunk.get_chunk_id(), chunk);
+
+    auto& heapLock = it->second->first;
+    auto& heap = it->second->second;
+
+    std::lock_guard<std::mutex> l(heapLock); 
+    heap.emplace(chunk);
+}
+
+//select_chunk will only be called in homestore when creating a shard.
+csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const homestore::blk_alloc_hints& hint) {
+    auto& chunkIdHint = hint.chunk_id_hint;
+    if(chunkIdHint.has_value()) {
+        LOGINFO("should not allocated a chunk with exiting chunk_id {} in hint!", chunkIdHint.value());
+        return nullptr;
+    }
+
+    //shardid -> chunkid map is maintained by ShardManager
+    //pg_id->pdev_id map is maintained by PgManager
+    //chunselector will not take care of the two maps for now.
+    uint32_t pdevID = 0;
+    auto& pdevIdHint = hint.pdev_id_hint;
+    if(!pdevIdHint.has_value()) {
+        // this is the first shard of this pg, select a pdev with the most available blocks for it
+        auto&& it = std::max_element(m_pdev_avalable_blk_map.begin(), m_pdev_avalable_blk_map.end(),
+        [](const std::pair<uint32_t, uint32_t>& lhs, const std::pair<uint32_t, uint32_t>& rhs) { 
+            return lhs.second < rhs.second; }
+        );
+        if(it == m_pdev_avalable_blk_map.end()) {
+            LOGINFO("No pdev found for new pg");
+            return nullptr;
+        }
+        pdevID = it->first;
+    }else{
+        pdevID = pdevIdHint.value();
+    } 
+
+    auto it = m_pdev_heap_map.find(pdevID);
+    if (it == m_pdev_heap_map.end()) {
+        LOGINFO("No pdev found for pdev {}", pdevID);
+        return nullptr;
+    }
+
+    const auto& vchunk = [it = std::move(it), pdevID]() {
+        auto& heapLock = it->second->first;
+        auto& heap = it->second->second;
+        std::lock_guard<std::mutex> l(heapLock);
+        if(heap.empty()) return VChunk(nullptr);
+        const auto& vchunk = heap.top();
+        heap.pop();
+        return vchunk;
+    }();
+
+    if(vchunk.get_internal_chunk()){
+        auto&& avalableBlkCounter = m_pdev_avalable_blk_map.find(pdevID)->second;
```

Comment:
type m_pdev_avalable_blk_map -> m_pdev_available_blk_map

Reply:
I also don't get this:
```
auto&& ...
```
What does this mean? I think the result is...a copy? It's not a reference, so it's moving the temporary? Can we just make it `auto`?

Reply:
actually, `auto&&` here is universal reference, and `->second` is a lvalue， so here it will be deduced to a lvalue reference。to make it more straightforward, we can change it to `auto&`

Reply:
:+1: